[OMPIRBuilder] Don't use invalid debug loc in declare mapper function.

abidh opened the pull request:
This is similar to https://github.com/llvm/llvm-project/pull/147950 but for declare mapper function.

abidh committed:
[OMPIRBuilder] Don't use invalid debug loc in declare mapper function.

This is similar to https://github.com/llvm/llvm-project/pull/147950 but
for declare mapper function.

abidh commented:
> Are you still interested in merging this? It looks fine to me -- I've applied this change on top of main, and I'm currently building it. I can run a few tests on it once built just to make sure everything is ok with it.

Yes, I would like to merge it. I think it is a useful cleanup. Please let me know if you testing finds anything.

abidh committed:
Tighten the testcase.

abidh commented:
Hi @skatrak @TIFitis @tblah

I am going through the remaining places in `OpenMPIRBuilder` where the insertion point and the debug location go out of sync. This PR fixes one of them: the declare mapper function has no `DISubprogram` of its own, but the instructions generated into it inherit the debug location of the function being compiled. With flang -g, every instruction in .omp_mapper.* ends up pointing at the enclosing subroutine. 

Apologies for the noise on the PR. I deleted the branch by accident while force-pushing a rebase, which closed it. It is restored and reopened; the only changes since @timsmith78 looked at it are a rebase onto main and a tightened test. A review would be appreciated.

TIFitis commented on mlir/test/Target/LLVMIR/omptarget-debug-declare-mapper.mlir:
```suggestion
// CHECK: ret void{{$}}
```
`CHECK: ret void` would still match `ret void, !dbg ....`.

TIFitis reviewed (approved):
LGTM with minor nit.

abidh committed:
Anchor the ret check to the end of the line.

Without it, "ret void" also matches a ret that carries a debug location,
which is exactly what this test is meant to catch: CHECK-NOT only covers
the text before the match, so a "!dbg" trailing the ret slips through.

timsmith78 commented:
LGTM.  I tried some of our OMP mapper tests and did not see any issues.
